morphia/src/main/java/org/mongodb/morphia/DatastoreImpl.java

```diff
@@ -1144,12 +1151,37 @@ protected void ensureIndex(final MappedClass mc, final DBCollection dbColl, fina
         }
 
         LOG.debug(format("Creating index for %s with keys:%s and opts:%s", dbColl.getName(), keys, opts));
-        dbColl.createIndex(keys, opts);
+        createMongoIndex(dbColl, keys, opts);
     }
 
-    protected void ensureIndex(final DBCollection dbColl, final DBObject keys, final DBObject options) {
+    protected void ensureIndex(final DBCollection dbColl, final DBObject keys, final BasicDBObject options) {
         LOG.debug(format("Creating index for %s with keys:%s and opts:%s", dbColl.getName(), keys, options));
-        dbColl.createIndex(keys, options);
+        createMongoIndex(dbColl, keys, options);
+    }
+
+    private void createMongoIndex(final DBCollection dbColl, final DBObject keys, final DBObject options) {
+        try {
+            dbColl.createIndex(keys, options);
+        } catch (MongoCommandException ex) {
+            final BasicDBObject basicOptions = (BasicDBObject) options;
+            if (basicOptions.getBoolean("dropOutdated", false)) {
+                final BsonDocument errorDocument = ex.getResponse();
+                if (errorDocument.getInt32("code").intValue() == 85) {
+                    String indexName;
+                    if (options.containsKey("name")) {
+                        indexName = basicOptions.getString("name");
+                    } else {
+                        final BsonString errorMessage = errorDocument.getString("errmsg");
```

Comment:
While I assume the tests show that this approach works, adding any logic based on the string of an error message is extremely brittle, and not necessarily correct (for example error messages vary between versions and could also potentially be translated). In an ideal world, we would rely on the error codes, but unfortunately we're not able to do that - error codes are not well documented, and are not consistent between MongoDB versions.